src/pubsub/ua_pubsub_connection.c

```diff
@@ -298,4 +303,234 @@ UA_PubSubConnection_regist(UA_Server *server, UA_NodeId *connectionIdentifier,
     return retval;
 }
 
+
+/* Connection State */
+
+static UA_StatusCode
+setPubSubState_disable(UA_Server *server,
+                                      UA_PubSubConnection *connection,
+                                      UA_StatusCode cause) {
+    UA_WriterGroup *wg;
+    UA_ReaderGroup *rg;
+    switch (connection->state){
+        case UA_PUBSUBSTATE_DISABLED:
+            break;
+        case UA_PUBSUBSTATE_PAUSED:
+            break;
+        case UA_PUBSUBSTATE_PREOPERATIONAL:
+        case UA_PUBSUBSTATE_OPERATIONAL:
+            LIST_FOREACH(wg, &connection->writerGroups, listEntry){
+                UA_WriterGroup_setPubSubState(server, wg, UA_PUBSUBSTATE_DISABLED,
+                                                cause);
+            }
+            LIST_FOREACH(rg, &connection->readerGroups, listEntry){
+                UA_ReaderGroup_setPubSubState(server, rg, UA_PUBSUBSTATE_DISABLED,
+                                                cause);
+            }
+
+            break;
+        case UA_PUBSUBSTATE_ERROR:
+            break;
+        default:
+            UA_LOG_WARNING_CONNECTION(&server->config.logger, connection,
+                                        "Received unknown PubSub state!");
+    }
+    connection->state = UA_PUBSUBSTATE_DISABLED;
+    return UA_STATUSCODE_GOOD;
+}
+
+static UA_StatusCode
+setPubSubState_paused(UA_Server *server,
+                                     UA_PubSubConnection *connection,
+                                     UA_StatusCode cause) {
+    UA_LOG_DEBUG_CONNECTION(&server->config.logger, connection,
+                             "PubSub state paused is unsupported at the moment!");
+    (void)cause;
+    switch (connection->state) {
+        case UA_PUBSUBSTATE_DISABLED:
+            break;
+        case UA_PUBSUBSTATE_PAUSED:
+            break;
+        case UA_PUBSUBSTATE_PREOPERATIONAL:
+            break;
+        case UA_PUBSUBSTATE_OPERATIONAL:
+            break;
+        case UA_PUBSUBSTATE_ERROR:
+            break;
+        default:
+            UA_LOG_WARNING_CONNECTION(&server->config.logger, connection,
+                                        "Received unknown PubSub state!");
+    }
+    return UA_STATUSCODE_BADNOTSUPPORTED;
+}
+
+static UA_StatusCode
+setPubSubState_preoperational(UA_Server *server,
+                                            UA_PubSubConnection *connection,
+                                            UA_StatusCode cause) {
+    switch(connection->state) {
+        case UA_PUBSUBSTATE_DISABLED:
+        case UA_PUBSUBSTATE_PAUSED:
+            connection->state = UA_PUBSUBSTATE_PREOPERATIONAL;
+            return UA_STATUSCODE_GOOD;
+        case UA_PUBSUBSTATE_PREOPERATIONAL:
+            break;
+        case UA_PUBSUBSTATE_OPERATIONAL:
+            return UA_STATUSCODE_GOOD;
```

Comment:
If you go from operational to pre-operational, wouldn't you have to close the underlying connection?
(e.g. MQTT...)